test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryTest.cs

```diff
@@ -362,6 +475,59 @@ public void ModelSetter_SameType_UpdatesModelExplorer()
             Assert.NotSame(originalExplorer, viewData.ModelExplorer);
         }
 
+        [Theory]
+        [InlineData(null)]
+        [InlineData(23)]
+        public void CopyConstructor_DoesNotChangeMetadata_WhenValueCompatible(int? model)
+        {
+            // Arrange
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var metadata = metadataProvider.GetMetadataForType(typeof(int?));
+            var explorer = new ModelExplorer(metadataProvider, metadata, model: -48);
+            var source = new ViewDataDictionary(metadataProvider)
+            {
+                ModelExplorer = explorer,
+            };
+
+            // Act
+            var viewData = new ViewDataDictionary(source, model);
+
+            // Assert
+            Assert.NotNull(viewData.ModelExplorer);
+            Assert.NotSame(explorer, viewData.ModelExplorer);
+            Assert.Same(metadata, viewData.ModelMetadata);
```

Comment:
This assertion fails w/o the `src` changes. `viewData.ModelMetadata.ModelType` is `typeof(object)` when `model==null` and `typeof(int)` when `model==23`.